### PR TITLE
Mode-dependent RISC-V disassembly

### PR DESCRIPTION
**Pull request by arichardson**
Constantly reading the wrong instructions was sufficiently annoying so I tried to fix this. I'm not sure if this approach is the best way of doing this, but it appears to work.
If this seems okay I'll add support for the remaining mode-dependent instructions.

With this change mode-dependent instructions such as AUIPCC/CLC/etc. will
be disassembled as the capmode version.
In order to do determine the mode we are in right now we pass an additional
flag to the disassembler. If this flag is not passed (e.g. in the objdump
case), we fall back to querying the ELF header to check for EF_RISCV_CAPMODE.

Note: this is currently not quite correct for GDB: there is no cached value
for capmode so we use is_cheriabi instead (which is generally the same but
is not necessarily true).

TODO: A follow-up change should set the value of capmode based on PCC.flags
when the current PCC is available instead of statically based on the ELF
header of the main program.

### CODE REVIEW

File: opcodes/riscv-dis.c [hunk 1]
```diff
@@ -484,12 +485,30 @@ riscv_disassemble_insn (bfd_vma memaddr, insn_t word, disassemble_info *info)
...
 	}
-
+      /*
+       * Set the value of capmode based on the flags that GDB passed or fall
+       * back to the ELF header if not available.
+       */
+      if (info->flags & RISCV_OPCODE_FLAG_CAPMODE_ENABLED)
+	in_capmode = 1;
+      else if (info->flags & RISCV_OPCODE_FLAG_CAPMODE_DISABLED)
+	in_capmode = 0;
```
Comment (arichardson) [inline opcodes/riscv-dis.c:495]: Seems like github's web UI can't handle the (rather odd) GDB indentation rules. I just copied the xlen indentation from above.

**Comment by jrtc27** [comment]
IMO this belongs more in the `isa` field as that's really a predicates field...

**Comment by arichardson** [comment]
> IMO this belongs more in the `isa` field as that's really a predicates field...

```
struct riscv_opcode
{
  /* The name of the instruction.  */
  const char *name;
  /* The requirement of xlen for the instruction, 0 if no requirement.  */
  unsigned xlen_requirement;
  /* An array of ISA subset name (I, M, A, F, D, Xextension), must ended
     with a NULL pointer sential.  */
  const char *subset[MAX_SUBSET_NUM];
  /* A string describing the arguments for this instruction.  */
  const char *args;
  /* The basic opcode for the instruction.  When assembling, this
     opcode is modified by the arguments to produce the actual opcode
     that is used.  If pinfo is INSN_MACRO, then this is 0.  */
  insn_t match;
  /* If pinfo is not INSN_MACRO, then this is a bit mask for the
     relevant portions of the opcode when disassembling.  If the
     actual opcode anded with the match field equals the opcode field,
     then we have found the correct instruction.  If pinfo is
     INSN_MACRO, then this field is the macro identifier.  */
  insn_t mask;
  /* A function to determine if a word corresponds to this instruction.
     Usually, this computes ((word & mask) == match).  */
  int (*match_func) (const struct riscv_opcode *op, insn_t word);
  /* For a macro, this is INSN_MACRO.  Otherwise, it is a collection
     of bits describing the instruction, notably any relevant hazard
     information.  */
  unsigned long pinfo;
};
```

**Comment by arichardson** [comment]
Unless you mean `subset`? But then that's a string comparison. Not sure it matters since GDB's disassembly is already stupidly slow but I don't think we need to slow it down even more.

**Comment by jrtc27** [comment]
Oh, the comments in riscv-opc.c still call it `isa`, but yeah. Who knows why it's not an enum. It's been refactored to an `enum riscv_insn_class` upstream though, but now a single one, not a list, so that's going to be really ugly even with non-mode-dependent CHERI instructions. I do wonder about the RISC-V maintainers sometimes.

**Comment by bsdjhb** [comment]
I ended up re-implementing this from scratch in the `cheri-12` branch.

**Pull request closed without merging**